### PR TITLE
Macgeargear/sucu 66 user update profile

### PR DESCRIPTION
**Pull request by macgeargear**
### Type of Change

- [x] New features
- [ ] Bug fixes
- [ ] Breaking changes
- [ ] Refactor
- [ ] Documentation updates
- [ ] Other (please specify):

## Description
- update user profile (admin method)

## Screenshots or GIFs (if applicable)
**Existing Database**
Screenshot 2024-09-14 at 2 57 18 PM

**mock user**
```go
req := &dtos.UserDTO{
    ID: "0000000001",
}
apperr := h.userUsecase.UpdateProfile(req, &updateProfile)
```

**update profile requests**
Screenshot 2024-09-14 at 2 55 04 PM

**updated user**
Screenshot 2024-09-14 at 2 57 59 PM

### CODE REVIEW

File: cmd/server/fiber_server.go [hunk 1]
```diff
@@ -120,6 +120,7 @@ func (s *FiberHttpServer) initUserRouter(router fiber.Router, httpHandler handle
 
 	userRouter.Get("/", httpHandler.User().GetAllUsers)
 	userRouter.Post("/", httpHandler.User().CreateUser)
+	userRouter.Put("/", httpHandler.User().UpdateProfile)
```
Comment (wiraphatys) [inline cmd/server/fiber_server.go]: ใส่ middleware isLogin ไปด้วยงับ ไม่งั้นใช้ c.Locals ของ fiber ctx ไม่ได้นะ

File: internal/domain/usecases/user_usecase.go [hunk 2]
```diff
@@ -39,7 +39,7 @@ func (u *userUsecase) GetUserByID(req *dtos.UserDTO, userID string) (*dtos.UserD
 }
 
 func (u *userUsecase) CreateUser(req *dtos.UserDTO, createUserDTO *dtos.CreateUserDTO) *apperror.AppError {
-	role, err := utils.GetRole(createUserDTO.Role)
+	role, err := utils.GetRole(req.Role)
```
Comment (wiraphatys) [inline internal/domain/usecases/user_usecase.go:42]: ขอบคุณมากคับ เมื่อวานตอนรีวิวมองข้ามจุดนี้55555555

File: internal/interface/repositories/user_repository.go [hunk 3]
```diff
@@ -30,14 +30,11 @@ func (r *userRepository) FindUserByID(ID string) (*entities.User, error) {
 }
 
 func (r *userRepository) InsertUser(user *entities.User) error {
-	if err := r.db.Create(user).Error; err != nil {
-		return err
-	}
-	return nil
+	return r.db.Create(user).Error
```
Comment (wiraphatys) [inline internal/interface/repositories/user_repository.go:33]: ประหยัดบรรทัดมากงับ

File: cmd/server/fiber_server.go (hunk 1 above)
Comment (wiraphatys) [reply cmd/server/fiber_server.go]: แล้วก็อยากให้ใช้ PATCH method คับ เพราะเราไม่ได้อัพเดตทั้ง object

Comment (wiraphatys) [reply cmd/server/fiber_server.go]: ปล. ใส่ middleware ทั้ง POST,  PATCH เลยนะะ

Comment (wiraphatys) [reply cmd/server/fiber_server.go]: @macgeargear

Comment (macgeargear) [reply cmd/server/fiber_server.go]: > ใส่ middleware isLogin ไปด้วยงับ ไม่งั้นใช้ c.Locals ของ fiber ctx ไม่ได้นะ

okk อันนี้ลืมจิง 5555

**Comment by bookpanda** [comment]
Please also make sure if POST body only have some fields e.g. only firstname, lastname and password stays the same.

File: internal/domain/usecases/user_usecase.go [hunk 4]
```diff
@@ -90,5 +90,10 @@ func (u *userUsecase) DeleteUserByID(req *dtos.UserDTO, userID string) *apperror
 // admin method
 
 func (u *userUsecase) UpdateProfile(req *dtos.UserDTO, updateUserDTO *dtos.UpdateUserDTO) *apperror.AppError {
+	err := u.userRepository.UpdateUserByID(req.ID, updateUserDTO)
+	if err != nil {
+		u.logger.Named("UpdateProfile").Error(constant.ErrUpdateUserByID, zap.String("userID", req.ID), zap.Error(err))
+		return apperror.BadRequestError(constant.ErrUpdateUserByID)
+	}
```
Comment (bookpanda) [inline internal/domain/usecases/user_usecase.go]: Let's also hash password here

**Comment by bookpanda** [comment]
let's also gen update swagger
```
swag init -d ./internal/interface/handlers -g ../../../cmd/main.go -o ./docs -md ./docs/markdown --parseDependency --parseInternal
```
maybe put it in a Makefile? @wiraphatys

File: internal/domain/usecases/user_usecase.go (hunk 4 above)
Comment (macgeargear) [reply internal/domain/usecases/user_usecase.go]: ลืมเลย แหะๆ😂

**Comment by macgeargear** [comment]
> Please also make sure if POST body only have some fields e.g. only firstname, lastname and password stays the same.

I add some field validation logic here
- I use the `map[string]interface{}` to avoid sending zero-value fields from the DTO, ensuring only intended changes are made.
- Returns an error if no fields are provided.

**Comment by macgeargear** [comment]
> let's also gen update swagger
> 
> ```
> swag init -d ./internal/interface/handlers -g ../../../cmd/main.go -o ./docs -md ./docs/markdown --parseDependency --parseInternal
> ```
> 
> maybe put it in a Makefile? @wiraphatys

Good idea krub, I already add it

**Comment by macgeargear** [comment]
I think we could add Docker Compose and a Dockerfile so that everyone can easily set up the environment without needing to configure everything themselves.
@bookpanda @wiraphatys what do you guys think?

**Comment by wiraphatys** [comment]
actually I already provided all of backend env on linear [https://linear.app/sucu/issue/SUCU-58/environment-config] and we have only 2 files there are docker-compose.yml and .env

**Comment by wiraphatys** [comment]
sorry but did I told all of you guys that our env is on linear @macgeargear

**Review by wiraphatys — APPROVED** [review]

**Comment by macgeargear** [comment]
> sorry but did I told all of you guys that our env is on linear @macgeargear 

ท่ดทีพึ่งเหนว่ามี555

**Comment by wiraphatys** [comment]
wait for @bookpanda ngubb

**Comment by bookpanda** [comment]
lgtm

> > Please also make sure if POST body only have some fields e.g. only firstname, lastname and password stays the same.
> 
> I add some field validation logic here
> 
> * I use the `map[string]interface{}` to avoid sending zero-value fields from the DTO, ensuring only intended changes are made.
> * Returns an error if no fields are provided.

i see

**Comment by bookpanda** [comment]
lgtm

**Review by bookpanda — APPROVED** [review]

**Comment by wiraphatys** [comment]
let merge by urself kub bro @macgeargear

**Pull request merged**